Node Auto Arrangement in GraphEdit/VisualScript/VisualShader

theoway opened the pull request:
**This pull request is part of this GSoC '21 project:** https://summerofcode.withgoogle.com/projects/#6101762760507392

- Closes https://github.com/godotengine/godot-proposals/issues/1253.

**Tasks completed**:

- Added button to call `GraphEdit::arrange_nodes()`
- Implemented `GraphEdit::arrange_nodes()` (*_See the proposal for reference_)
- Enabled undo/redo
- Tested in VisualScript/VisualShaders Editor
- Refactoring done

**Future work**:
- Handle edge case when no node is selected
- Minor style changes
- Add `SpinBox` for horizontal & vertical gap control


**Current Status**: Merged.

theoway commented:
@fire This is the draft PR. I'll update the description with an intro and a task list. Also with each step, I'll be adding visuals so that progress can be monitored

charlesmlamb commented:
This is 'stunning' feature, can't wait <3

Hopefully, it'll be in a Godot 4 alpha, etc . .

Well, 9.8/10 <3

fire commented on scene/gui/graph_edit.cpp:
```diff
@@ -1666,14 +1666,81 @@ HashMap<int, Vector<StringName>> GraphEdit::layering(const Set<StringName> &v, c
+				if (j == k) {
+					cross.set(q, crossings);
+				}
+				else if (j < k) {
```
I prefer reversed conditions and I believe this doesn't follow the coding standard and the cicd will block the merge.

```
if (!var) {
    continue;
}

fire replied:
There's a new line after } else. I'm referring to this.

theoway replied:
Will fix it right away. Forgot to ran document formatter before pushing! My bad...

fire commented:
Making it ready for review so cicd runs, but still in progress.

pycbouh commented:
> Making it ready for review so cicd runs, but still in progress.

Why do we need to make CI run if it's unfinished? We better not overburden it with pointless actions.

theoway commented:
https://user-images.githubusercontent.com/41969735/125192806-3f832400-e267-11eb-8600-6cf8245d88e8.mp4

@fire 
The `GraphNodes` are now placed in blocks, with each block having straight connections/edges. Here's a demo. After selecting the nodes and hitting the shortcut, nodes are rearranged.
**Next update:** Next would be to remove the extra space between the blocks. After this, we'll have a compact layout!

DavidCambre commented:
Just wanted to put my name in for testing when you are ready.
Thumb this comment if you want to be added to test-crew.

DavidCambre commented:
I like that you may just keep trying and stop when you like it.
![Auto Arrangement](https://user-images.githubusercontent.com/20573784/126788090-67947887-f4c5-400d-9041-e18e4fcb41df.gif)

theoway commented:
@DavidCambre Yes, you are likely to get a different layout every time, since the algorithm uses `randomization` to minimize the crossings between the connections.

theoway commented:
https://user-images.githubusercontent.com/41969735/126897282-1618d2db-042f-4648-aa65-dcf74473b8c2.mp4

The extra space between blocks has been reduced to get a compact layout and horizontal gaps between nodes has been increased for better readability.

charlesmlamb commented:
Looking at latest video, wow . . 10 / 10 feature . . .

Two things, that I'm thinking . . .

It's not enough to ' order ' the boxes, I think, the goal is to make the graph readable, not just align the boxes . . That's maybe an impossible goal, for an algorhithm . . .

In a sense, in video above, it was more readable before, bec. the boxes didn't have lines running under nodes, in a sense, we also need an anti - spaghetti arranger, that might take inventing an entirely new algorhithm . . .

So, in the above video, what I'm thinking is, yes, at first, it's done an amazing job, boxes are placed, more neatly, there's structure . . . But, the real question is, had it become more readable, also . . . .

Anyway, looking at above video, I'd say the feature is done, bas. ready for, the engine, AMAZING work . . .

What will define, if I use the feature, is not just if it makes the boxes stack and, so on . . . It'll be, is graph more readable, can I ' see ' logic arrows and, are the boxes stacked, in a ' logical ' way, so it reads well <3

Amazing work, already . .

theoway commented:
@charlesmlamb Thanks for the feedback! You have said valid points about readability, also for which, in the next commit, I'm adding a few more things. First would be more space between nodes horizontally, so that the users can have more clarity when figuring out the connections, adding to readability. Also, I'll be adding GUI button/elements to adjust horizontal & vertical gaps, so that the users can customize the whole layout as per their preference..

charlesmlamb commented:
Ok, do your thing, as a long - term user of Visual script, readability is a factor but, it might not be possible to ' code ' that . . .

Also, ordering the boxes, so they're visually pleasing or, a ' joy ' to read, well, it's NOT important but, I do that . .

I think the core feature, should just be ' ordered ' and, later, after getting some reviews and, thinking about it, any-way the latest update, I was thinking, feature is done . . Do, what you feel is best and, for the engine, project . . .

Also, I tend to make some adjustments, after using the algorhithm, so it's not important . . . .

I'd say, right now, what you've got, is a good feature, it's basically ' done ' . . .

Hoping you'll consider these things and, find a good solution, that ' makes ' sense . . . 

Thx <3 . . .

charlesmlamb commented:
Well, we want Visual script to be as strong as possible, a ' happy ' feature . .

Rest, is to each user <3

charlesmlamb commented:
Ok, here's how I do it, to get best readability, after making some scripts . .

First, place _ready() at top, then _process(), _physicsprocess(), then new functions . . .

Wanted to make them, in a long line, not vertically, but it sort-of leads, to a ton, of panning . .

<img width="1920" alt="Skærmbillede (1194)" src="https://user-images.githubusercontent.com/83387915/127655103-0da8b8b0-292c-492f-854f-430993f68128.png">

I try, to make boxes fit in squares, triangles, or polygons . . .

<img width="1920" alt="Skærmbillede (1342) - Kopi" src="https://user-images.githubusercontent.com/83387915/127655230-9586577b-5341-41a4-8a91-0c3159233d32.png">

<img width="870" alt="Skærmbillede (1889)" src="https://user-images.githubusercontent.com/83387915/127655325-28b7a092-40d1-48d1-aa05-9f256da2762c.png">

In the end, that's how I arrange scripts, after some experience . . There could be many other ways, it sort makes it a bit more ' ordered ' and, there are ' rules ', for how the boxes are placed . . .

After some experiences, wanted to place all functions, in one line, horizontally, but it sort of led, to panning . . So, it sadly bec. necessary, to place them vertically . . . :OO . . .

<img width="1920" alt="Skærmbillede (1907)" src="https://user-images.githubusercontent.com/83387915/127655551-49938d61-550f-4a1b-a3be-2eaa9d4561d6.png">

So, in the end, scripts tend to look, like they fit inside a big square . . . :OO

<img width="1920" alt="Skærmbillede (1341)" src="https://user-images.githubusercontent.com/83387915/127655670-e019b4e0-a9b7-4894-a28a-ba92b889bc5e.png">

Well, it leads to less panning, since the script, is one big square . .

I'm not coding it, it's just my experience, atm, how to arrange scripts . . But, I'm not coding and, well it's important, that the coder, makes the calls . . . :OO 

Thx <3

charlesmlamb commented:
Last post, here's a video about how I do it, atm . .

> https://www.youtube.com/watch?v=NogysQ44g6c

It's not the ' best ' way, don't really know . . . But, it sort leads to ' least ' panning and, makes all script readable, bec. they all look like that, so you don't have to think, in a new script, how it ' works ' . . .

You're the coder, so do your thing, it's your ' work ', system, just some ideas, don't really know, what's best way, thanks . . . .

theoway commented:
@charlesmlamb Well, after going through your examples, the end result of my algorithm is quite similar to the way you arrange nodes. By the time this project is finished, everything will be more polished

fire commented on scene/gui/graph_edit.cpp:
```diff
@@ -1646,6 +1646,585 @@ HBoxContainer *GraphEdit::get_zoom_hbox() {
 	return zoom_hb;
 }
 
+int GraphEdit::set_operations(SET_OPERATIONS e, Set<StringName> &U, const Set<StringName> &V) {
```
A common godot convention is to use p_ for parameter and r_ for reference (changes input). Please update.

theoway replied:
I just have to make a few more additions. After that, I'll change the variables all together

fire commented on scene/gui/graph_edit.cpp:
```diff
@@ -1646,6 +1646,585 @@ HBoxContainer *GraphEdit::get_zoom_hbox() {
+	}
+
+	//Print layers here
+	//Check if vectors of hashmap are being written
```
When the pr is ready to be submitted, make tests or delete the commented code because they cause confusion.

theoway replied:
Sure!

fire commented on scene/gui/graph_edit.cpp:
```diff
@@ -1646,6 +1646,585 @@ HBoxContainer *GraphEdit::get_zoom_hbox() {
+		}
+
+		if (incoming.from != StringName()) {
+			GraphNode *gfrom = Object::cast_to<GraphNode>(node_names[incoming.from]);
```
Godot uses `snake_case`. You are using `gfrom` as your variable names. Please change style.

pycbouh replied:
It's completely fine to name local variables like that. It can even be `gf` or something.

fire commented on scene/gui/graph_edit.cpp:
```diff
@@ -1646,6 +1646,585 @@ HBoxContainer *GraphEdit::get_zoom_hbox() {
+		new_positions.set(E->get(), cal_pos);
+		block_number++;
+	}
+	//Now finding positions of nodes inside a block
```
Coding style.

> //Now finding positions of nodes inside a block

> // Now finding positions of nodes inside a block.

theoway replied:
Oops! My bad. I'll add comments as per the guidelines

fire commented on scene/gui/graph_edit.cpp:
```diff
@@ -1646,6 +1646,585 @@ HBoxContainer *GraphEdit::get_zoom_hbox() {
+
+	//Moving nodes to new positions
+	emit_signal("begin_node_move");
+	for (const Set<StringName>::Element *E = selected_nodes.front(); E; E = E->next()) {
```
We are able to use `for (const StringName &e : selected_nodes)` now I think. Need to check.

theoway replied:
If it works, then I'll change it all the possible places.

charlesmlamb commented:
Ok, hoping for a good feature, thanks . . .

Gallilus commented:
Wanted to test this script. it crashes on import.
So I hope you are able to use it in the future

![image](https://user-images.githubusercontent.com/20573784/128381895-64e1d435-0c93-498e-873a-a01c5cc36b1b.png)


[mvp_basic_functions (5).zip](https://github.com/godotengine/godot/files/6939990/mvp_basic_functions.5.zip)

theoway commented:
@Gallilus You said crashes on import. Can it mean that the script is corrupted? Or did the crash occur after using the feature on your script. Anyways, I'll test your script and see how it goes.

Gallilus commented:
Created #51291 to document the workaround I found for courted files. So this one should work.
[auto_arrange (2).zip](https://github.com/godotengine/godot/files/6940617/auto_arrange.2.zip)

I do not know how far you are in the algorithm. So my apologies if I'm early, as you see not ideal yet.
![image](https://user-images.githubusercontent.com/20573784/128396201-673deeb5-4a37-48bb-99ce-71c1f9698c86.png)

theoway commented:
@charlesmlamb Thanks to your script, I found out another bug. Will fix it right away, then run fixed code on your script and share the results here :)

theoway commented:
Ran it on the last portion on your visualscript. The results with the fixed code look like _this_. I think the layout is pretty neat and readable. Will make the commit in sometime. Let me know your view on this updated layout :)

![Screenshot (286)](https://user-images.githubusercontent.com/41969735/128468070-a4d435a0-966a-4d58-a610-3e055e90b135.png)

theoway commented:
Fixed the code. Now you won't see large vertical offsets like before. I ran the code on a portion of this script. Results are below.  IMO, it's looking good, although there can be some improvements. For example: The nodes that are stacked vertically in the beginning of the picture can be spread across the empty space. But I guess that's a different story. As far as the algorithm is concerned, this is the best that you can get out of it. Although in the future, I can try new variations to make it look even more better(*I've already started finding new research papers lol!) 
Also, feel free to perform more tests and report any bugs here. I'm looking forward to getting it merged in a week or so ;)
![Screenshot (287)](https://user-images.githubusercontent.com/41969735/128479521-9bc01192-929e-431b-86b5-575cd8f9e194.png)

Gallilus commented:
Il keep testing it.
Indeed I would love to see the data nodes staggered as the sequence nodes.
Until then I think the advised workflow would be. Add a portion of logic, auto arrange, continue, repeat.

theoway commented:
> Il keep testing it.
> Indeed I would love to see the data nodes staggered as the sequence nodes.
> Until then I think the advised workflow would be. Add a portion of logic, auto arrange, continue, repeat.

Indeed, I think users can greatly benefit from this, following your suggested workflow. For now, I'm gonna focus on GUI and refactoring so that it gets merged soon! :)

theoway commented:
![Screenshot (289)](https://user-images.githubusercontent.com/41969735/128590962-af9ce56d-535a-4e4f-a5ae-ef6f78fcdd30.png)

@Gallilus Looks much better if applied on a portion, then repeating the whole procedure. One can see the difference of applying it to a small portion and large portion respectively. I think one has more control over the layout by working on small chunks of nodes. Lemme know your thoughts 👍🏻

theoway commented:
https://user-images.githubusercontent.com/41969735/128676228-4a14c590-b3cd-46dd-abb0-964ef86d334c.mp4

Now the nodes can be arranged by pressing the button as shown in the video.(_*Will change the icon later, this is just for testing purpose! Contributions for the icon are welcome!_ )
Now to add `spinbox` to adjust the gaps.

theoway commented:
@Calinou Could you add `GSoC` tag to this PR, as it's this year's GSoC project?

charlesmlamb commented:
This is, looking good . . Wow . .

10 / 10 . .

WesCook commented:
It may already work this way, but the button "Arrange selected nodes" could simply be "Arrange nodes" when no nodes are selected (ie. it acts on all of them).  That would save having to manually select everything before use.

The feature is looking very nice so far.

theoway commented:
@WesCook Yeah, I think that's a valid point. In the next update, will add that. Even if the user accidentally arranges all the nodes while what they actually wanted to arrange some specific nodes, they can simply undo it.

theoway committed:
Automatic arrangement of nodes in VisualScript/VisualShaders editors

This PR and commit adds the functionality to arrange nodes in VisualScript/VisualShader editor. The layout generated by this 
feature is compact, with minimum crossings between connections
& uniform horizontal & vertical gaps between the nodes. 

This work has been sponsored by GSoC '21.

Full list of additions/changes:
• Added arrange_nodes() method in GraphEdit module.
    • This method computes new positions for all the selected
      nodes by forming blocks and compressing them.
      The nodes are moved to these new positions. 
    • Adding this method to GraphEdit makes it available for 
      use in VisualScript/VisualShaders editors and its other
      subclasses. 
• Button with an icon has been added to call arrange_nodes() in GraphEdit. 
    • This button is inherited by VisualScript/VisualShaders editors
       to invoke the method.
• Undo/redo is functional with this method.
    • By using signals in arrange_nodes(), position changes are registered 
       in undo/redo stack of the subclass that is using the method. 
• Metadata of the method has been updated in ClassDB
• Method description has been added to class reference of GraphEdit

fire commented on scene/gui/graph_edit.h:
```diff
@@ -116,6 +116,8 @@ class GraphEdit : public Control {
 
 	Button *minimap_button;
 
+	Button *layout_button;
```
Todo for future pr. initialize to nullptr.

fire commented on scene/gui/graph_edit.cpp:
```diff
@@ -1646,6 +1647,500 @@ HBoxContainer *GraphEdit::get_zoom_hbox() {
+		} while (u != E->get());
+	}
+
+	//Compute horizontal co-ordinates individually for layers to get uniform gap
```
We can collect fixes into a later pr to fix comments. Minor style correct.

fire reviewed (approved):
Thanks!

RPicster commented:
I know this is closed for a while - but would a backport be a lot of work?
I'm on a project where this could be a BIG game changer.

theoway commented:
> I know this is closed for a while - but would a backport be a lot of work? I'm on a project where this could be a BIG game changer.

@RPicster Depends on which Godot version you want it to work on? I don't think it'll be much trouble for 3.x

RPicster commented:
That's fantastic news! 😁 Thank you so much!
I'm always on stable for work projects.

theoway commented:
@RPicster I'll make a draft PR for `3.x`. Will refer it here so that you and others can test it.

RPicster commented:
> @RPicster I'll make a draft PR for `3.x`. Will refer it here so that you and others can test it.

Fantastic! Thank you!

RPicster commented:
@theoway 

Sorry for the bump without contributing any new things.... but did you have a chance to look into the backport? :)

theoway commented:
> @theoway 
> 
> Sorry for the bump without contributing any new things.... but did you have a chance to look into the backport? :)
> 

I didn't get the chance to work on it. Lately, I have not been able to contribute to Godot. As soon as I get back, this backport is my highest priority. 
Meanwhile, if anyone else wants to do it, please feel free to ping me here 😄
